codex/blockexchange/engine/advertiser.nim

```diff
@@ -0,0 +1,173 @@
+    advertiseTasks*: seq[Future[void]]                           # Advertise tasks
+
+    advertiseLocalStoreLoopSleep: Duration                                 # Advertise loop sleep
+    inFlightAdvReqs*: Table[Cid, Future[void]]                   # Inflight advertise requests
+
+proc addCidToQueue(b: Advertiser, cid: Cid) {.async.} =
+  if cid notin b.advertiseQueue:
+    await b.advertiseQueue.put(cid)
+    trace "Advertising", cid
+
+proc advertiseBlock(b: Advertiser, cid: Cid) {.async.} =
+  without isM =? cid.isManifest, err:
+    warn "Unable to determine if cid is manifest"
+    return
+
+  if isM:
+    without blk =? await b.localStore.getBlock(cid), err:
+      error "Error retrieving manifest block", cid, err = err.msg
+      return
+
+    without manifest =? Manifest.decode(blk), err:
+      error "Unable to decode as manifest", err = err.msg
+      return
+
+    # announce manifest cid and tree cid
+    await b.addCidToQueue(cid)
+    await b.addCidToQueue(manifest.treeCid)
+
+proc advertiseLocalStoreLoop(b: Advertiser) {.async.} =
+  while b.advertiserRunning:
+    if cids =? await b.localStore.listBlocks(blockType = BlockType.Manifest):
+      trace "Advertiser begins iterating blocks..."
+      for c in cids:
+        if cid =? await c:
+          await b.advertiseBlock(cid)
+      trace "Advertiser iterating blocks finished."
+
+    await sleepAsync(b.advertiseLocalStoreLoopSleep)
+
+  info "Exiting advertise task loop"
+
+proc processQueueLoop(b: Advertiser) {.async.} =
+  while b.advertiserRunning:
+    try:
+      let
+        cid = await b.advertiseQueue.get()
+
+      if cid in b.inFlightAdvReqs:
+        continue
+
+      try:
+        let
+          request = b.discovery.provide(cid)
+
+        b.inFlightAdvReqs[cid] = request
+        codexInflightAdvertise.set(b.inFlightAdvReqs.len.int64)
+        await request
+
+      finally:
+        b.inFlightAdvReqs.del(cid)
+        codexInflightAdvertise.set(b.inFlightAdvReqs.len.int64)
+    except CancelledError:
+      trace "Advertise task cancelled"
+      return
+    except CatchableError as exc:
+      warn "Exception in advertise task runner", exc = exc.msg
+
+  info "Exiting advertise task runner"
+
+proc start*(b: Advertiser) {.async.} =
+  ## Start the advertiser
+  ##
+
+  trace "Advertiser start"
+
+  proc onBlock(cid: Cid) {.async.} = 
+    await b.advertiseBlock(cid)
+
+  b.localStore.onBlockStored = onBlock.some
+
+  if b.advertiserRunning:
+    warn "Starting advertiser twice"
+    return
+
+  b.advertiserRunning = true
+  for i in 0..<b.concurrentAdvReqs:
+    b.advertiseTasks.add(processQueueLoop(b))
+
+  b.advertiseLocalStoreLoop = advertiseLocalStoreLoop(b)
+
+proc stop*(b: Advertiser) {.async.} =
+  ## Stop the advertiser
+  ##
+
+  trace "Advertiser stop"
+  if not b.advertiserRunning:
+    warn "Stopping advertiser without starting it"
+    return
+
+  b.advertiserRunning = false
+  for task in b.advertiseTasks:
```

Comment:
Ah yes. There are two ways in which tasks get added to the queue:
1. The local-store-advertising loop loops.
2. The local-store onBlockStored callback fires.

setting the `advertiserRunning = false` will stop the advertising loop, and so no new tasks will come from there. But I did miss the callback. So I'm adding `b.localStore.onBlockStored = CidCallback.none`. This will prevent new tasks from being added while we're in the process of shutting down.

Good catch!

Reply:
:red_circle: OK so I think this can still fail to provide what you want to guarantee. You need to synchronize on loop termination; i.e. you need to move the `if not b.advertiseLocalStoreLoop.isNil and not b.advertiseLocalStoreLoop.finished` block before this, otherwise it can happen that:

1. the loop (`processQueueLoop`) checks that `advertiserRunning` is true and starts an iteration;
2. the loop blocks on `await listBlocks`;
3. someone calls your shutdown procedure, and it runs the whole task cancellation logic;
4. the loop awakens from `await listBlocks` and enqueues another task, which at this point won't get cancelled.

Of course one could wonder if this guarantee needs to be strict, but if it doesn't then I guess there's no point in cancelling the tasks at all.